onnx/test/shape_inference_test.py

```diff
@@ -3455,6 +3455,39 @@ def test_trilu_lower_one(self):  # type: () -> None
             [],)
         self._assert_inferred(graph, [make_tensor_value_info('y', TensorProto.INT32, (3, 1, 5))])  # type: ignore
 
+    def test_batch_norm_train(self):  # type: () -> None
+        graph = self._make_graph(
+            [('x', TensorProto.FLOAT, (3, 4, 5, 6, 7)),
+             ('scale', TensorProto.FLOAT, (4,)),
+             ('b', TensorProto.FLOAT, (4,)),
+             ('mean', TensorProto.FLOAT, (4,)),
+             ('var', TensorProto.FLOAT, (4,)),
+             ('training_mode', TensorProto.BOOL, ())],
+            [make_node('BatchNormalization', ['x', 'scale', 'b', 'mean', 'var', 'training_mode'],
+                       ['out', 'output_mean', 'output_var', 'saved_mean', 'saved_var'])],
```

Comment:
How about adding more tests to test robustness of the shape inference method :

1) Shapes with dim_params (along with dim_values)
2) Shapes with no dim_values and no dim_params ?
3) Missing shapes for some inputs

Reply:
What is the difference between 2 and 3?

Reply:
2 -> shapes are present for inputs but they contain no dim_value or dim_param.
3 -> Shapes are missing altogether for certain inputs.

Does this make sense ?